source/adaptive_load/adaptive_load_controller_impl.h

```diff
@@ -0,0 +1,80 @@
+#include "envoy/common/time.h"
+
+#include "nighthawk/adaptive_load/adaptive_load_controller.h"
+#include "nighthawk/adaptive_load/metrics_evaluator.h"
+#include "nighthawk/adaptive_load/session_spec_proto_helper.h"
+#include "nighthawk/adaptive_load/step_controller.h"
+#include "nighthawk/common/nighthawk_service_client.h"
+
+namespace Nighthawk {
+
+class AdaptiveLoadControllerImpl : public AdaptiveLoadController {
+public:
+  /**
+   * Constructs an implementation of the adaptive load controller main loop that relies on logic in
+   * several helper objects. Through helpers, it performs Nighthawk Service benchmarks, obtains
+   * metrics from MetricsPlugins, scores the results, and consults a StepController plugin to
+   * determine the next load and detect convergence. All plugins are specified through the
+   * AdaptiveLoadSessionSpec proto.
+   *
+   * Usage:
+   *
+   *   AdaptiveLoadControllerImpl controller(
+   *       NighthawkServiceClientImpl(),
+   *       MetricsEvaluatorImpl(),
+   *       AdaptiveLoadSessionSpecProtoHelperImpl(),
+   *       Envoy::Event::RealTimeSystem()); // NO_CHECK_FORMAT(real_time))
+   *   AdaptiveLoadSessionSpec spec;
+   *   // (set spec fields here)
+   *   StatusOr<AdaptiveLoadSessionOutput> output =
+   *       controller.PerformAdaptiveLoadSession(&nighthawk_service_stub, spec);
+   *
+   * @param nighthawk_service_client A helper that executes Nighthawk Service benchmarks given a
+   * gRPC stub.
```

Comment:
What is the reason behing seperating the client from the stub?

Reply:
This is for historical reasons: We had a helper that made it easier to call the stub, then moved the helper into its own class for testing purposes.

I'm not sure what will happen with this, but we may add a more convenient Nighthawk Service client wrapper object that contains the stub, and then take this wrapper instead. Or the stub might be the best thing for the public API to take.